pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the adjoint_metric_tensor.
+"""
+import warnings
+from pennylane import numpy as np
+
+import pennylane as qml
+
+
+def _get_generator(op):
+    """Reads out the generator and prefactor of an operation and converts
+    to matrix if necessary.
+
+    Args:
+        op (:class:`~.Operation`): Operation to obtain the generator of.
+    Returns:
+        array[float]: Generator matrix
+        float: Prefactor of the generator
+    """
+
+    generator, prefactor = op.generator
+    if not isinstance(generator, np.ndarray):
+        generator = generator.matrix
+    if op.inverse:
+        generator = generator.conj().T
+        prefactor *= -1
+
+    return generator, prefactor
+
+
+def _apply_any_operation(state, op, device, invert=False):
+    """Wrapper that allows to apply a variety of operations---or groups
+    of operations---to a state or to prepare a new state.
+    If ``invert=True``, this function makes sure not to alter the operations.
```

Comment:
what do you mean by alter?

Reply:
`_apply_any_operation` reverts the change to an operation that happens if we call `op.inv()`, namely the flipping of the `inverse` flag.
Otherwise, reusing the list of operations would introduce errors.